sdk/core/System.ClientModel/src/Pipeline/MessageLoggingPolicy.cs

```diff
@@ -0,0 +1,342 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// A <see cref="PipelinePolicy"/> used by a <see cref="ClientPipeline"/> to
+/// log request and response information.
+/// </summary>
+public class MessageLoggingPolicy : PipelinePolicy
+{
+    /// <summary>
+    /// The <see cref="ClientRetryPolicy"/> instance used by a default
+    /// <see cref="ClientPipeline"/>.
+    /// </summary>
+    public static MessageLoggingPolicy Default { get; } = new MessageLoggingPolicy();
+
+    private readonly bool _enableMessageContentLogging;
+    private readonly int _maxLength;
+    private readonly PipelineMessageLogger _messageLogger;
+    private readonly string _clientAssembly = typeof(MessageLoggingPolicy).Assembly.GetName().Name!;
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="MessageLoggingPolicy"/> class.
+    /// </summary>
+    /// <param name="options">The user-provided logging options object.</param>
+    public MessageLoggingPolicy(ClientLoggingOptions? options = default)
```

Comment:
TODO: If a policy takes in freezable options, should it freeze them?

Reply:
Good observation that with this sequencing, options can change after the policy is constructed if we pass them to policy constructors.  Let's think about this a bit in terms of the life cycles of the various components in this story.

Reply:
The freezing operation is idempotent, so there is no problem if we call it in both places.  I think the question is whether we want to keep the options un-frozen after a policy is created until the pipeline is created with the root options.  It would be good to look at the tradeoffs around that -- I think we can implement it to satisfy either design ... it is more a question of what is the right design.

Reply:
I assume it's unlikely that someone would change the options after passing them to the logging policy constructor and expect the constructor to respect that change. It doesn't really make sense at all. I'm just thinking about this from a correctness perspective irt freezing. I do think it makes more sense to only freeze options once the pipeline is created, and not in policies themselves. I guess in that case then policies should probably just hold onto options and wait to use them until process/process async is called. But idk that feels weird to me.

Reply:
My experience with client options and customer support says otherwise.   If they can mutate options after passing them into a client constructor, they get angry when the change isn't respected.

As an example, we clone the options in the messaging libraries to freeze values.   We still see the occasional issue about "why did my change not get picked up" from those who don't read the docs where we talk about that.

Reply:
@jsquire Interesting! That's definitely different from how I understand constructors and options types personally. But I'm thinking I will just hold onto the options instance and let the pipeline manage freezing.